File: Test_2/Strings/Longest_Substring_Without_Repeating_Characters/Efficient_1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int>mp;
        int i = 0;
        int j = 0;
        int res = 0;
        while(j < s.size()) {
            char currChar = s[j];
            if(mp.find(currChar) == mp.end()) {
                mp[currChar]++;
                res = max(res, j - i + 1);
            } else if(mp.find(currChar) != mp.end()) {
                while(mp.find(currChar) != mp.end()) {
                    mp[s[i]]--;
                    if(mp[s[i]] == 0) mp.erase(s[i]);
                    i++;
                }
                mp[s[j]]++;
                res = max(res, j - i + 1);
            }
            j++;
        }
        return res;
    }
};

File: Test_2/Strings/Longest_Substring_Without_Repeating_Characters/Efficient_1.java
// Problem Link -> https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer> mp = new HashMap<>();
        int res = 0;
        int low = 0;
        int high = 0;
        while(high < s.length()) {
            char currChar = s.charAt(high);
            if(mp.containsKey(currChar)) {
                 while(mp.containsKey(currChar)) {
                    char toRemove = s.charAt(low);
                    int currOccurences = mp.get(toRemove);
                    currOccurences--;
                    if(currOccurences == 0) {
                        mp.remove(toRemove);
                    } else {
                        mp.put(toRemove, currOccurences);
                    }
                    low++;
                }
                mp.put(currChar, 1);
                res = Math.max(res, high - low + 1);
                
            } else {
                mp.put(currChar,1);
                res = Math.max(res, high - low + 1);
            }
            high++;
        }
        return res;
    }
}